File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/CASMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.bridge

import com.cleveradssolutions.plugin.flutter.CASChannel
import com.cleveradssolutions.plugin.flutter.CASFlutterContext
import com.cleveradssolutions.plugin.flutter.success
import com.cleversolutions.ads.android.CAS
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/cas"

class CASMethodHandler(
    binding: FlutterPluginBinding,
    private val contextService: CASFlutterContext
) : CASChannel(binding, CHANNEL_NAME) {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getSDKVersion" -> getSDKVersion(result)
            "validateIntegration" -> validateIntegration(call, result)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun getSDKVersion(result: MethodChannel.Result) {
        result.success(CAS.getSDKVersion())
    }

    private fun validateIntegration(call: MethodCall, result: MethodChannel.Result) {
        val activity = contextService.getActivityOrError(call, result) ?: return
        CAS.validateIntegration(activity)
        result.success()
    }

}

File: ios/Classes/Bridge/CASMethodHandler.swift
//
//  CASMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

class CASMethodHandler: CASChannel {
    init(with registrar: FlutterPluginRegistrar) {
        super.init(with: registrar, on: "cas")
    }

    override func onMethodCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "getSDKVersion": getSDKVersion(result)
        case "validateIntegration": validateIntegration(result)
        default: super.onMethodCall(call, result)
        }
    }

    private func getSDKVersion(_ result: @escaping FlutterResult) {
        result(CAS.getSDKVersion())
    }

    private func validateIntegration(_ result: @escaping FlutterResult) {
        CAS.validateIntegration()
        result(nil)
    }
}
